azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAppRoleAuthenticationFilter.java

```diff
@@ -0,0 +1,86 @@
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.proc.BadJOSEException;
+import com.nimbusds.jwt.proc.BadJWTException;
+import com.nimbusds.oauth2.sdk.util.CollectionUtils;
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import net.minidev.json.JSONArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+public class AADAppRoleAuthenticationFilter extends OncePerRequestFilter {
+
+    private static final Logger log = LoggerFactory.getLogger(AADAppRoleAuthenticationFilter.class);
+
+    private static final String TOKEN_HEADER = "Authorization";
+    private static final String TOKEN_TYPE = "Bearer ";
+    private static final JSONArray DEFAULT_ROLE_CLAIM = new JSONArray().appendElement("USER");
+    private static final String ROLE_PREFIX = "ROLE_";
+
+    private UserPrincipalManager principalManager;
+
+    public AADAppRoleAuthenticationFilter(UserPrincipalManager principalManager) {
+        this.principalManager = principalManager;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
+        FilterChain filterChain) throws ServletException, IOException {
+
+        final String authHeader = request.getHeader(TOKEN_HEADER);
+
+        if (authHeader != null && authHeader.startsWith(TOKEN_TYPE)) {
+            try {
+                final String token = authHeader.replace(TOKEN_TYPE, "");
+
+                final UserPrincipal principal = principalManager.buildUserPrincipal(token);
+                final JSONArray roles = Optional.ofNullable((JSONArray) principal.getClaims().get("roles"))
+                    .filter(r -> !r.isEmpty())
+                    .orElse(DEFAULT_ROLE_CLAIM);
+                final Authentication authentication = new PreAuthenticatedAuthenticationToken(
+                    principal, null, rolesToGrantedAuthorities(roles));
+
+                authentication.setAuthenticated(true);
+                log.info("Request token verification success. {}", authentication);
+                SecurityContextHolder.getContext().setAuthentication(authentication);
+            } catch (BadJWTException ex) {
+                log.warn("Invalid JWT: " + ex.getMessage());
+                throw new ServletException(ex);
```

Comment:
For log message for exception, my suggestion may looks like this `throw new Exception("Invalid JWT: Expired xxxxx", ex);` which include some message to user, as well as the cause.

Reply:
I am not sure I understand what you mean. When this exception is catched, `ex.getMessage()` will contain either `Expired JWT` or `JWT before use time`. From my point of view, there is no additional information I can provide in order to make more sense out of this exception. Or do you mean the fact that here `throw new ServletException(ex);`  no message is specified and only available in the causedBy?

Reply:
I mean if there are some message here, we can get the hint from log instead of debugging or something like that `ex.getMessage()` to find the cause. It is not big issue here, just from the users view about the exception throw.